WordPress/Classes/Services/PostCoordinator+UploadActionUseCase.swift

```diff
@@ -0,0 +1,95 @@
+
+import Foundation
+
+extension PostCoordinator {
+    enum UploadAction {
+        /// Upload the post as is.
+        ///
+        /// For example, if the post was published locally, it will be published when the server receives it.
+        case upload
+        /// Upload a revision to the server.
+        case autoSave
+        case nothing
+    }
+
+    /// Provides business logic for automatic uploads.
+    final class UploadActionUseCase {
+        private static let allowedStatuses: [BasePost.Status] = [.draft, .publish]
+
+        /// Returns what action should be executed when we retry a failed upload.
+        ///
+        /// In some cases, we do not want to automatically upload a post if the user has not
+        /// given explicit confirmation. Users "confirm" automatic uploads by pressing the
+        /// Publish or Update button in the editor.
+        ///
+        /// If we do not receive a confirmation, which can happen if the editor crashed, we will
+        /// try to upload a revision instead.
+        func autoUploadAction(for post: AbstractPost) -> UploadAction {
+            guard post.isFailed,
+                let status = post.status,
+                UploadActionUseCase.allowedStatuses.contains(status),
+                !post.hasRemote() else {
+                return .nothing
+            }
+
+            if post.isLocalDraft || post.confirmedAutoUpload {
+                return .upload
+            } else {
+                // TODO This is currently not supported by PostCoordinator
+                return .autoSave
+            }
+        }
+
+        /// Returns true if the post will be automatically uploaded later and it can be canceled.
+        ///
+        /// This can be used to determine if the app should show the Cancel button in the Post List.
+        ///
+        /// - SeeAlso: autoUploadAction(for:)
+        func canCancelAutoUpload(of post: AbstractPost) -> Bool {
+            guard autoUploadAction(for: post) == .upload else {
+                return false
+            }
+
+            // Local drafts are always automatically uploaded
+            if post.isLocalDraft {
+                return false
+            } else {
+                return post.confirmedAutoUpload
+            }
+        }
+
+        /// Temporary method to support old _Retry_ upload functionality.
+        func canRetryUpload(of post: AbstractPost) -> Bool {
+            guard post.isFailed,
+                let status = post.status else {
+                return false
+            }
+
+            return !UploadActionUseCase.allowedStatuses.contains(status)
```

Comment:
Just want to see if I understand. If the post status is draft or publish we will hide the retry button?

Reply:
Yes. The Retry button will only be shown for those Post status types that we do not currently auto-upload like `publishedPrivate` and `scheduled`. We shouldn't need the Retry button anymore since they will be automatically uploaded.

I'm sorry. In my quest to go breadth-first, implementing the whole flow first for `.published` and `.draft` instead of everything all at once, the old features are still there and can be confusing.